### PR TITLE
Helm tweaks

### PR DESCRIPTION
**Pull request by sjpb**
This started out as non-functional changes - see commit messages for these.

There are a few functional changes though:

- It's [undocumented](https://github.com/kubernetes/kubernetes/issues/50345), but using `subpath` in `volumeMounts` is what is preventing config maps from being updated. So this changes it so `/etc/slurm/` is the volume (so the empty directory in the image gets overwritten. A helm upgrade will now update the configmap, and (it may take 10s of seconds) the slurm.conf file will be updated. An `scontrol update` will then load the changed config. Unfortunately there is no way of doing this with the slurmdbd due to the need to have `slurmdbd.conf` a) in the same directory as `slurm.conf` and b) owned by `slurm`, but that doesn't use too much of the configuration in slurm.conf anyway.

- Options for slurm daemons can now be set in the kubernetes resource template using the container args.

- Its possible to use something like this: `docker run --rm -it <container tag> /bin/bash` now to look at the container outside of Kubernetes. This is useful for e.g. seeing which files exist.

- Its possible to pass `debug` as the container arg (instead of e.g. `slurmd`) in a Kubernetes resource to just start munge, which allows you to shell into the running container and e.g. debug changes to the entrypoint script.

- Because nodes are defined in slurm.conf, they are defaulting https://slurm.schedmd.com/slurm.conf.html#OPT_CPUs 1x CPU. Which means launching jobs requiring > 1 task per node fails. I've modified slurm.conf to use the current cluster configuration of 4x CPUs, mostly as a reminder this needs fixing via Azimuth. The `RealMemory` parameter should also be set by Azimuth.

### CODE REVIEW

File: image/docker-entrypoint.sh [hunk 1]
```diff
@@ -78,37 +78,36 @@ then
...
-    for DIR in */;
-    do USER_TO_SET=$( echo $DIR | sed "s/.$//" ) && (chown -R $USER_TO_SET:$USER_TO_SET $USER_TO_SET || echo "Failed to take ownership of $USER_TO_SET") \
-     && (chmod 700 /home/$USER_TO_SET/.ssh || echo "Couldn't set permissions for .ssh directory for $USER_TO_SET") \
-     && (chmod 600 /home/$USER_TO_SET/.ssh/authorized_keys || echo "Couldn't set permissions for .ssh/authorized_keys for $USER_TO_SET");
+    pushd /home > /dev/null
+    for DIR in *
+        do
+        chown -R $DIR:$DIR $DIR || echo "Failed to change ownership of $DIR"
+        chmod 700 $DIR/.ssh || echo "Couldn't set permissions for .ssh/ directory of $DIR"
+        chmod 600 $DIR/.ssh/authorized_keys || echo "Couldn't set permissions for .ssh/authorized_keys for $USER_TO_SET"
```
Comment (wtripp180901) [inline image/docker-entrypoint.sh:98]: USER_TO_SET doesn't exist anymore, could cause problems if setting permissions fails

File: README.md [hunk 2]
```diff
@@ -132,11 +132,21 @@ Note: The mpirun script assumes you are running as user 'rocky'. If you are runn
...
+```
+
+The new Slurm configuration can then be read by running `scontrol reconfigure` as root inside a Slurm pod. The [slurm.conf documentation](https://slurm.schedmd.com/slurm.conf.html) notes that some changes require a restart of all daemons, which here requires redeploying the Slurm pods as described below.
+
+Changes to other configuration files (e.g. Munge key etc) require a redeploy of the appropriate pods.
+
+To redeploy pods use:
 ```console
-kubectl rollout restart deployment <deployment-names>
+kubectl rollout restart deployment <deployment-names ...>
```
Comment (wtripp180901) [inline README.md:147]:
````suggestion
kubectl rollout restart deployment <deployment-names ...>
```
for the `slurmdbd`, `login` and `mysql` pods and

```
kubectl rollout restart statefulset <statefulset-names ...>
```
for the `slurmd` and `slurmctld` pods
````

File: image/docker-entrypoint.sh (hunk 1 above)
Comment (sjpb) [reply image/docker-entrypoint.sh:98]: In what way?  We should still get the same error message, I think? What am I missing here?

**Review by wtripp180901 — CHANGES_REQUESTED** [review]
Tested and all works on my end, just a few tweaks and doc changes

Also when trying to run `scontrol reconfig` after mounting in a new slurm.conf, if I try to change properties like MaxNodeCount or NodeName, it fails with error
`slurm_reconfigure error: Zero Bytes were transmitted or received`
and loses the worker nodes until they are manually restarted. It still works okay with stuff like changing the debug level. Is this expected behaviour?

File: image/docker-entrypoint.sh (hunk 1 above)
Comment (wtripp180901) [reply image/docker-entrypoint.sh:98]: I don't think its defined anywhere else in the shell script anymore

Comment (sjpb) [reply image/docker-entrypoint.sh]: Gotcha good catch, thanks.

**Comment by sjpb** [comment]
> Tested and all works on my end, just a few tweaks and doc changes
> 
> Also when trying to run `scontrol reconfig` after mounting in a new slurm.conf, if I try to change properties like MaxNodeCount or NodeName, it fails with error `slurm_reconfigure error: Zero Bytes were transmitted or received` and loses the worker nodes until they are manually restarted. It still works okay with stuff like changing the debug level. Is this expected behaviour?



I think so. See https://slurm.schedmd.com/slurm.conf.html#OPT_MaxNodeCount:
> .... The slurmctld daemon must be restarted for changes to this parameter to take effect.

The `slurm.conf` docs similarly note any other parameters where this is required, and the `scontrol reconfigure` [docs](https://slurm.schedmd.com/scontrol.html) also say:

> Parameters requiring a restart will be noted in the [slurm.conf](https://slurm.schedmd.com/slurm.conf.html)(5) man page. The slurmctld daemon and all slurmd daemons must also be restarted if nodes are added to or removed from the cluster.

**Pull request merged**